### DISCUSSION TITLE
Valetudo 2022.02.0

### DISCUSSION THREAD

**Discussion by Hypfer** [Hypfer/Valetudo#1379] [Release Discussions]
<div align="center">
    <img src="https://raw.githubusercontent.com/Hypfer/Valetudo/master/assets/logo/valetudo_logo_with_name.svg" width="800" alt="valetudo">
    <p align="center"><h2>2022.02.0</h2></p>
</div>

Lots of UI changes and more polishing

## Quirks

This release adds the quirks concept, which shall be understood as a catch-all and/or staging area containing vendor-specific toggles that don't fit the generic abstraction that is Valetudo (yet).
This should make it fairly easy to quickly implement (some of the) exciting new vendor features without jeopardizing the architecture of Valetudo in the long run.

Here's an example taken from the Dreame Z10:

![image](https://user-images.githubusercontent.com/974410/151532264-ff6b848f-0c5a-4083-af92-d3bc29be77a0.png)

## Settings

Robot, Map and Connectivity settings have been reorganized/redone.

![image](https://user-images.githubusercontent.com/974410/151532305-e97a3534-3f8a-44bc-8d4d-dec5537d7c70.png)

![image](https://user-images.githubusercontent.com/974410/151532320-456cda6f-b5f7-4104-95f5-4b258d0b576b.png)

![image](https://user-images.githubusercontent.com/974410/151532335-d0a79b27-b1ae-4114-b5f2-e5f2311ed8e0.png)

Wi-Fi and NTP state display has been redone as well:

![image](https://user-images.githubusercontent.com/974410/151532430-07629c75-e112-4d8c-acb6-5138d0d6de3d.png)

![image](https://user-images.githubusercontent.com/974410/151532445-ba2b34b4-6b25-4a22-9a7b-558054405b3f.png)

## Quality of Life

A _lot_ of help text sections and dialogs have been added all around the application to make usage of Valetudo even easier.
They should also answer a lot of common support questions, so make sure to read them before asking questions.

Furthermore, all password fields have been updated to feature a plain-text-display toggle.

## Total Statistics

Because displaying three numbers is boring, @ccoors had the great idea of adding gamification the total statistics feature.
After some iteration on that idea, we've ended up with this:

![image](https://user-images.githubusercontent.com/974410/151532462-7843b526-9c81-4b7f-bffe-1165d7dd02c9.png)

There's also an overview of all achievements:

![image](https://user-images.githubusercontent.com/974410/151532488-d0cfc8d4-c477-4ed6-8d48-edcfe6d6bec7.png)

I'm very happy with how this turned out. Now we just need to think of more achievements.
If you have any ideas, feel free to leave them down in the comments.

## Nightly builds

There are now automated nightly builds, which you can install using the updater.

As this is meant for people willing to accept and capable of handling breakage, there is no UI toggle to switch to nightly builds.
To enable them, ssh into your robot and change the update provider to `github_nightly`.

## Misc

- Viomi consumables have been fixed by @adrez99
- @schinken fixed local access via IPv6

### Features

* **core:** WifiScanCapability ([a4fe895](https://github.com/Hypfer/Valetudo/commit/a4fe895c9944ffe8fae6e621a87e2aa70e9aecde))
* **ui:** Add general help ([ec298dc](https://github.com/Hypfer/Valetudo/commit/ec298dc2cf2d81ca40031dfae455abf490f9fcf8))
* **ui:** Add help for dnd, voice packs and the updater ([84c8c04](https://github.com/Hypfer/Valetudo/commit/84c8c04657778795cb7e7acd76bbef001465ba3a))
* **ui:** Add Map Management Help ([9d62c26](https://github.com/Hypfer/Valetudo/commit/9d62c26318ecda5d5a38e21333f56ac0ff1f70c6))
* **ui:** Add quirks help ([77183ed](https://github.com/Hypfer/Valetudo/commit/77183ed3e63c94d30afbefc6d48cab8bf946b9f2))
* **ui:** Add total statistics page ([#1330](https://github.com/Hypfer/Valetudo/issues/1330)) ([8fd0b2e](https://github.com/Hypfer/Valetudo/commit/8fd0b2eb3c59a5b949da8c9c6f0f4230886d48e6))
* **ui:** Allow display of Wi-Fi passwords entered ([89624eb](https://github.com/Hypfer/Valetudo/commit/89624eb5b6c2b8091ab66821ebf1c3f82f119316))
* **ui:** Allow newlines in confirmation dialog ([df2ee4f](https://github.com/Hypfer/Valetudo/commit/df2ee4f0ac0e209ac537a77705a860c5161f756b))
* **ui:** Another achievement ([0ffbef4](https://github.com/Hypfer/Valetudo/commit/0ffbef4f3d64e1c50ae6663c277ef33d9c2e80ce))
* **ui:** Extend map management help ([e1514bc](https://github.com/Hypfer/Valetudo/commit/e1514bcc59f8c0022aef588b58b10e2cc05c1ef9))
* ValetudoWifiStatus  may contain a bssid ([86fe2f2](https://github.com/Hypfer/Valetudo/commit/86fe2f207c2f050ca7351c21bd736de4f2d5d27b))
* **ui:** Allow display of entered passwords for mqtt and basic auth ([4e75531](https://github.com/Hypfer/Valetudo/commit/4e7553100abb4dba247fbb036a6539499c619b89))
* **ui:** Disable edit map tap interaction while not docked to reduce confusion ([b68e444](https://github.com/Hypfer/Valetudo/commit/b68e44428b3c2669cae147c8c4f21bbfc9bc016d))
* **ui:** Improve achievement badges ([7eea13e](https://github.com/Hypfer/Valetudo/commit/7eea13e84950d01c044bc69ef385b33908ee20df))
* **ui:** Improve NTP client state display ([0a265a7](https://github.com/Hypfer/Valetudo/commit/0a265a7dce5468e2948a077bd7e6f364212a285b))
* **ui:** More achievements ([892017e](https://github.com/Hypfer/Valetudo/commit/892017ea38c16eee3ac6bc52d8ff5bde76e87ab1))
* **ui:** Move map-related features to map management page ([ea45f01](https://github.com/Hypfer/Valetudo/commit/ea45f01b6a597f43d6000318500936702c89e397))
* **ui:** Move Wi-Fi settings to connectivity ([dff909c](https://github.com/Hypfer/Valetudo/commit/dff909cae688b2106cc526475247cf0659251e46))
* **ui:** Provide an overview of all achievements ([381e3cd](https://github.com/Hypfer/Valetudo/commit/381e3cd4dab807b85d8066ca4aa9f8628fe599d6))
* **ui:** Rename about page to system information ([ad1dd16](https://github.com/Hypfer/Valetudo/commit/ad1dd1608dc0e2668914d7ab7a2eb68cb6978163))
* **ui:** Restructure connectivity settings ([bfe33e4](https://github.com/Hypfer/Valetudo/commit/bfe33e4d4d929741a1a3191559ae84343dbe9a61))
* **ui:** Restructure robot settings ([5c71be4](https://github.com/Hypfer/Valetudo/commit/5c71be44eb4a9f30674af550e230590b203ea82d))
* **ui:** Restructure routers and add about page ([e3a1f13](https://github.com/Hypfer/Valetudo/commit/e3a1f1303bb8693aea0a8ea7ffb32d8fd658110a))
* **updater:** Add nightly builds ([d1a0d91](https://github.com/Hypfer/Valetudo/commit/d1a0d919a6fd7df93031bdb7da469b1b61cdbc13))
* **vendor.dreame:** Add auto empty interval quirk ([1b6205a](https://github.com/Hypfer/Valetudo/commit/1b6205a467304bcd4c9ce32aa28721fe0fd377f5))
* **vendor.roborock:** Add quirks ([7fc584a](https://github.com/Hypfer/Valetudo/commit/7fc584af029efbe2a7b8e063fb9659afa43d0955))
* **vendor.viomi:** Add quirks ([#1369](https://github.com/Hypfer/Valetudo/issues/1369)) ([b369e4d](https://github.com/Hypfer/Valetudo/commit/b369e4d7a253bd668f354f179a878fd8ce9c0717))
* QuirksCapability ([79deeb1](https://github.com/Hypfer/Valetudo/commit/79deeb14eb855e294f0b9d3197151e8af5cc03ce))
* **webserver:** Allow IPv6 requests from own network ([#1342](https://github.com/Hypfer/Valetudo/issues/1342)) ([165deff](https://github.com/Hypfer/Valetudo/commit/165deff8e033b45ec1e507d4678833dbca8b22d4))

### Bug Fixes

* **mqtt:** Fix consumables not being polled by the autorefresh ([58d6267](https://github.com/Hypfer/Valetudo/commit/58d6267e3008c6637725ed7f531f07d182b01ddf)), closes [#1355](https://github.com/Hypfer/Valetudo/issues/1355)
* **networkadvertisement:** Attempt to catch issues of the ssdp lib ([ee3154b](https://github.com/Hypfer/Valetudo/commit/ee3154bfdb6c5caadc22382e0e48e9a42a2df202))
* **ui:** Allow refetching when there are zero quirks ([7f3ed2b](https://github.com/Hypfer/Valetudo/commit/7f3ed2b8b688ecf9f26436209625dda35e654b62))
* **ui:** Allow user-selection of wifi IP addresses ([33c9470](https://github.com/Hypfer/Valetudo/commit/33c9470162ba4fe6ae2bbf5ba2de26ea08f1d47f))
* **ui:** Allow user-selection of wifi IP addresses... again ([2bdcea4](https://github.com/Hypfer/Valetudo/commit/2bdcea4701b8828dfbcdde79aac12823ece07295))
* **ui:** Disable updater buttons when busy ([f82ae1a](https://github.com/Hypfer/Valetudo/commit/f82ae1a0a114c3255c43d2700cab54a9f87067ff))
* **ui:** Don't display updater warning if state is busy ([4d5c781](https://github.com/Hypfer/Valetudo/commit/4d5c7811b74c3a61f3eebaf1064bb18f6a1810a0))
* **ui:** Don't provide the wifi configuration route if there is no wifi configuration capability ([82e954e](https://github.com/Hypfer/Valetudo/commit/82e954e2d072dcf7888aeef7eb732b78a73311f4))
* **ui:** Fix about title ([67a9a05](https://github.com/Hypfer/Valetudo/commit/67a9a051c7d2071099a57a3a6dbf1e4fa101ca31))
* **ui:** Fix display of text in conjunction with controls in Map Management ([1cc76c5](https://github.com/Hypfer/Valetudo/commit/1cc76c55f3c27be5f03e5a7cbfd6cc184ad40675))
* **ui:** Fix logviewer timestamp display for chrome ([56f5ba0](https://github.com/Hypfer/Valetudo/commit/56f5ba08a9bc864eb20df9b17f109c0b1a89226e))
* **ui:** Fix map management not using unique keys for all options ([2305b52](https://github.com/Hypfer/Valetudo/commit/2305b528be28351e3e02f6197ca88135dc95e64c))
* **ui:** Fix menu drawer scrollbars ([4a6c9fa](https://github.com/Hypfer/Valetudo/commit/4a6c9faffcefba07277372a9172aa6f899d795ad))
* **ui:** fix minor typo ([#1341](https://github.com/Hypfer/Valetudo/issues/1341)) ([77e07ec](https://github.com/Hypfer/Valetudo/commit/77e07ec04bb32407176ff184e706a4229cdffced))
* **ui:** Properly implement newlines in ConfirmationDialog ([b8f8eb0](https://github.com/Hypfer/Valetudo/commit/b8f8eb0163b8b3c13b5a2e1bab27837b8ff90811))
* **ui:** Sort total statistics data points ([a568c03](https://github.com/Hypfer/Valetudo/commit/a568c03afa0521779f40c4933935b5315a81bd24))
* **ui:** use relativ paths to allow hosting in a subdirectory .. again ([3007dbb](https://github.com/Hypfer/Valetudo/commit/3007dbb1460596591199ece1d974c3b4e1db58e3))
* **updater:** Add missing return statement ([f5e7451](https://github.com/Hypfer/Valetudo/commit/f5e7451a5877815a8871cdccdbf997f09b4b78ca))
* **vendor.viomi:** Fix Viomi consumables ([#1367](https://github.com/Hypfer/Valetudo/issues/1367)) ([ccfe175](https://github.com/Hypfer/Valetudo/commit/ccfe175886ac41ff58f900de54657967a8f17337))
* **webserver:** Fix missing doctype for error pages ([3d48611](https://github.com/Hypfer/Valetudo/commit/3d486114830eb69ba2f414aa67fdf4494cfefe3c))
* **webserver:** Fix some endpoints never returning anything ([32c0f6a](https://github.com/Hypfer/Valetudo/commit/32c0f6ae922f71a8fd7ac5b035b9f5fffd1cdc8f))

<hr /><em>This discussion was created from the release <a href='https://github.com/Hypfer/Valetudo/releases/tag/2022.02.0'>Valetudo 2022.02.0</a>.</em>

**Comment by sky321** [comment]
Thank you very much for another beautiful release :-)

**Comment by StefanSchoof** [comment]
For me the Updater will not Install this version:
```
Error while downloading release binary TypeError [ERR_INVALID_ARG_TYPE]: The "url" argument must be of type string. Received undefined
    at new NodeError (node:internal/errors:371:5)
    at validateString (node:internal/validators:119:11)
    at Url.parse (node:url:169:3)
    at Object.urlParse [as parse] (node:url:156:13)
    at dispatchHttpRequest (/snapshot/Valetudo/node_modules/axios/lib/adapters/http.js:118:22)
    at new Promise (<anonymous>)
    at httpAdapter (/snapshot/Valetudo/node_modules/axios/lib/adapters/http.js:48:10)
    at dispatchRequest (/snapshot/Valetudo/node_modules/axios/lib/core/dispatchRequest.js:58:10)
    at Axios.request (/snapshot/Valetudo/node_modules/axios/lib/core/Axios.js:108:15)
    at Axios.<computed> [as get] (/snapshot/Valetudo/node_modules/axios/lib/core/Axios.js:129:17)
    at Function.wrap [as get] (/snapshot/Valetudo/node_modules/axios/lib/helpers/bind.js:9:15)
    at Updater.download (/snapshot/Valetudo/backend/lib/updater/Updater.js:136:50)
    at Updater.triggerDownload (/snapshot/Valetudo/backend/lib/updater/Updater.js:89:14)
    at /snapshot/Valetudo/backend/lib/webserver/UpdaterRouter.js:37:42
    at Layer.handle [as handle_request] (/snapshot/Valetudo/node_modules/express/lib/router/layer.js:95:5)
    at next (/snapshot/Valetudo/node_modules/express/lib/router/route.js:137:13) {
  code: 'ERR_INVALID_ARG_TYPE'
}
```
I am on 2022.01.0 with my S5.
![Screenshot_20220130-173159](https://user-images.githubusercontent.com/4662023/151708224-90bd3023-0691-49d3-a593-4234a3ea6351.png)

Is there anything I can do to get the Updater to work or do I need to build a new image?

**Reply by Hypfer** [reply]
#1380

**Reply by StefanSchoof** [reply]
Thanks.

**Comment by gschmidl** [comment]
Achievement idea: running the battery down; getting the robot stuck repeatedly; mopping; remote controlling over a certain distance; cleaning every day for a week; not cleaning for a week.

**Comment by dmlemos** [comment]
Some ideas for achievements:
* How much full robot rotations are equivalent to each planet - https://sos.noaa.gov/catalog/datasets/planet-rotations/
* Birthday commemoration - starting from the first clean. 1, 3, 5 years
* Cleaned an area equivalent to the [British Museum](https://en.wikipedia.org/wiki/British_museum)
* Bowling! - bumped 100 times into obstacles
* Delivery driver - went to 365 go to locations. Equivalent to a year of deliveries

**Reply by rimrul** [reply]
I don't think we have a way to detect the robot bumping into things or rotating.

**Comment by KayGundhardt** [comment]
Hey,

after updating to 2022.02.0 I can't´t get the robots statistic-data anymore. Error message is 

```
2022-02-03T21:48:58.487Z
Warn
Error while fetching statistics Error: Received invalid response
    at RoborockTotalStatisticsCapability.getStatistics (/snapshot/Valetudo/backend/lib/robots/roborock/capabilities/RoborockTotalStatisticsCapability.js:15:19)
    at async /snapshot/Valetudo/backend/lib/webserver/capabilityRouters/StatisticsCapabilityRouter.js:8:26
```

Any help would be appreciated! 

Thank you!


Kay

**Reply by Hypfer** [reply]
Please set the loglevel to trace, navigate to the total statistics again and then check the log for the response for `get_clean_summary`

**Reply by Hypfer** [reply]
d1926a9b33dfdbdf53f8b3b53ef0ee695c53c867

Should be fixed in the next release

**Comment by KayGundhardt** [comment]
Thank you for help! 
The log shows:


```
2022-02-12T21:02:39.923Z
Debug
>>> cloud: {"method":"get_clean_summary","params":[],"id":8041}
```
![9289DF19-8620-401D-A857-D704DA6E54F0](https://user-images.githubusercontent.com/69385168/153728566-efec30f4-8231-491e-9d59-e5ae10b30cd0.jpeg)

**Reply by Hypfer** [reply]
We need the response to that. Either check the messages around that timestamp or search for one of those message IDs (8062, 8063, 8064, 8065)

**Reply by KayGundhardt** [reply]
Ah, okay - sorry! 

Hope this one is right?

Debug
>>> cloud: {"method":"get_consumable","params":[],"id":8260}
2022-02-12T21:32:49.556Z
Debug
<<< cloud: {"id":8260,"result":[{"main_brush_work_time":171220,"side_brush_work_time":377954,"filter_work_time":171220,"filter_element_work_time":0,"sensor_dirty_time":37305,"dust_collection_work_times":0}],"exe_time":2}
2022-02-12T21:32:49.647Z

**Reply by Hypfer** [reply]
That's the response for get_consumable. Not get_clean_summary

**Reply by KayGundhardt** [reply]
I´m really a jerk - sorry 🙈

I really hope I won't mess up again and this one is right:


```
2022-02-12T21:00:48.364Z
Debug
>>> cloud: {"method":"get_clean_summary","params":[],"id":8021}
2022-02-12T21:00:48.375Z
Debug
<<< cloud: {"id":8021,"result":{"clean_time":1095667,"clean_area":12458112500,"clean_count":969,"dust_collection_count":0,"records":[1644678276,1644677834,1644624000,1644539995,1644537599,1644451197,1644435272,1644381610,1644381098,1644367157,1644364798,1644278400,1644194363,1644191999,1644153423,1644107980,1644105601,1644019198,1643975018,1643974546]},"exe_time":3}
```

**Reply by Hypfer** [reply]
Yep that's the correct one. So which model of Robot and which FW version is it?

**Comment by KayGundhardt** [comment]
😎 Puh!
Roborock S7 (FW1398,11/21) 

I hope you mean the dustbuilded one?

**Comment by KayGundhardt** [comment]
Cool - THANK YOU :)
